### PR TITLE
[BD-21] Deprecate waffle namespaces

### PR DESCRIPTION
**Pull request by regisb**
**Description:** Deprecate WaffleFlagNamespace and WaffleSwitchNamespace. All related features and methods are displaced to the WaffleFlag/WaffleSwitch classes.

**JIRA:** https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation

**Reviewers:**
- [x] @robrap 

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

This PR depends on https://github.com/edx/edx-toggles/pull/98

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @regisb! I've created [BLENDED-650](https://openedx.atlassian.net/browse/BLENDED-650) to keep track of it in Jira.   More details are on the [BD-21](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1596358943/BD-21+Toggles+Settings+Documentation) project page.


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by regisb** [comment]
Following our conversation @robrap this is ready for review.

**Comment by robrap** [comment]
@regisb: FYI: I am going to let the dust settle on the earlier PR first. Please update relevant comments (Changelog, PR Description, etc.) that details the backward incompatible change on the interfaces of WaffleFlag and WaffleSwitch.  For example, dropping the namespace from the constructor, and possibly mentioning the properties/methods that are going away, etc.

File: CHANGELOG.rst
Comment (robrap) [inline CHANGELOG.rst:26]: I think the term "deprecated" is a little confusing from the standpoint of the new class.  How about?
```
    * The `WaffleSwitch.switch_name` attribute no longer exists.  Switches should only ever be referred to using their fully namespaced names.  If you need the non-namespaced name, it must be parsed from the namespaced name.
```

File: CHANGELOG.rst
Comment (robrap) [inline CHANGELOG.rst:32]: Similar to switch:
```
  * The `WaffleFlag.flag_name` attribute no longer exists.  Flags should only ever be referred to using their fully namespaced names.  If you need the non-namespaced name, it must be parsed from the namespaced name.
```

File: CHANGELOG.rst
Comment (robrap) [inline CHANGELOG.rst:34]: Again, instead of "deprecated" maybe something like:
```
      * The `WaffleFlag.waffle_namespace` attribute no longer exists, since there is no longer a separate namespace object.
```

File: edx_toggles/toggles/__future__.py
Comment (robrap) [inline edx_toggles/toggles/__future__.py:6]: Why were these ever added to `__futures__`?  Let's not add these back in.  We could mentioned this in the changelog if we want to be safe with it, but I'll leave that up to you.

File: edx_toggles/toggles/__future__.py
Comment (robrap) [inline edx_toggles/toggles/__future__.py:6]: Can we just import the 4 classes explicitly?  Why are we changing this to a * import?  Do the imports need to change at all?

File: edx_toggles/toggles/__init__.py
Comment (robrap) [inline edx_toggles/toggles/__init__.py:6]: I noted in Slack that we might need an earlier change where `LegacyWaffleFlag` becomes a real class to get accurate custom attributes.  If that is the case, this will get updated.

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/legacy.py:7]: This should be:
```
from edx_toggles.toggles.waffle import LegacyWaffleSwitch, LegacyWaffleSwitchNamespace
```

File: edx_toggles/toggles/internal/waffle/legacy.py
Comment (robrap) [inline edx_toggles/toggles/internal/waffle/legacy.py:18]:
```
from edx_toggles.toggles.waffle import LegacyWaffleFlag, LegacyWaffleFlagNamespace
```

**Review by robrap — COMMENTED** [review]
Thanks @regisb.  I must just be super picky. :)

File: CHANGELOG.rst
Comment (regisb) [reply CHANGELOG.rst:18]: It is not strictly required but it improves the readability.

File: edx_toggles/toggles/__future__.py
Comment (regisb) [reply edx_toggles/toggles/__future__.py:6]: It was just convenient to import all classes from a single module. This is not a breaking change, as we now `import *` from `toggles`.

File: edx_toggles/toggles/__future__.py
Comment (regisb) [reply edx_toggles/toggles/__future__.py:6]: I think that the `import *` makes it clearer which module needs to be imported.

File: edx_toggles/toggles/__init__.py
Comment (regisb) [reply edx_toggles/toggles/__init__.py:6]: > FYI: I don’t think I can tell if we have stopped using the current WaffleFlag and are only using LegacyWaffleFlag (or futures), because both are coming from the legacy module with the class WaffleFlag when monitoring. (Apologies if I am missing something.)
> Proposal:
> 
>* Make LegacyWaffleFlag (and the other Legacy classes) a real class that subclasses the existing class, also in the legacy module.
> * Update the module based custom attributes to be statically defined custom attributes that let you know which is being imported. For example, we could set deprecated_current_waffle_class in the following legacy module classes: WaffleFlag, WaffleSwitch, WaffleFlagNamespace, and WaffleSwitchNamespace, still with a value including the class and toggle name. We could set deprecated_legacy_waffle_class in the new legacy module classes: LegacyWaffleFlag, LegacyWaffleSwitch, etc.
> * We would want to ensure that deprecated_current_waffle_class had no data before publishing edx-toggles 2.0.0. Note: If I don’t see any data, I might consider temporarily importing something from the wrong place, just to prove to myself that I am querying the data correctly.

If I understand correctly, we want to:

1. check which apps have statements such as `from edx_toggles.toggles import WaffleSwitch, WaffleFlag`.
2. replace these statements by `from edx_toggles.toggles import LegacyWaffleSwitch, LegacyWaffleFlag`

This is currently not possible because the class name indicated by `LegacyWaffleFlag.__class__` is "WaffleFlag".

Thus, a simple fix would be to write in edx-toggles==1.2.1 (to be released):

```
class LegacyWaffleFlag(WaffleFlag): pass
class LegacyWaffleSwitch(WaffleSwitch): pass
```

As far as I understand, it is unnecessary to modify the set_attribute calls. We would just need to check for custom "edx_toggles.toggles.waffle.legacy.WaffleFlag" and "edx_toggles.toggles.waffle.legacy.WaffleSwitch" custom attributes.

Comment (robrap) [reply edx_toggles/toggles/__init__.py:6]: Thanks, I think you are correct as long as the two sets of classes are defined in two different modules. As defined, we won’t see all issues, but just the last issue per transaction. It is ok if one issue masks another issue until the first is fixed, but not ok if a non-issue masks an issue and we ne we learn about it.

File: CHANGELOG.rst
Comment (robrap) [inline CHANGELOG.rst:17]: Nit: You could use double backtick Around class names for consistency.

File: CHANGELOG.rst
Comment (robrap) [inline CHANGELOG.rst:28]: Nit: Some lines are using single backtick instead of double backtick around class and method names.

File: edx_toggles/toggles/__future__.py
Comment (robrap) [reply edx_toggles/toggles/__future__.py:6]: Maybe I’m missing something, but doesn’t this make the Legacy classes available from futures? Seems odd to me. In any case, this is temporary and it seems like you’ll be relying on search/replace when futures goes away, so whatever makes you happy. 🙂

File: edx_toggles/toggles/__init__.py
Comment (robrap) [reply edx_toggles/toggles/__init__.py:6]: Still awaiting response for separating these two sets of classes under different modules.

**Review by robrap — COMMENTED** [review]
Very close. Thanks. Reminder that this is blocked adjusting the custom attributes and ensuring monitoring is happy before merging/releasing.

File: edx_toggles/toggles/__init__.py
Comment (regisb) [reply edx_toggles/toggles/__init__.py:6]: Here is the monitoring PR: https://github.com/edx/edx-toggles/pull/98

File: edx_toggles/toggles/__future__.py
Comment (robrap) [reply edx_toggles/toggles/__future__.py:6]: @regisb: I'll assume this continues to make you happy and my comments don't change anything, so feel free to mark resolved.

Comment (regisb) [reply edx_toggles/toggles/__future__.py:6]: :ok_hand:

File: CHANGELOG.rst
Comment (robrap) [inline CHANGELOG.rst:33]: Nit: Another example of single backticks.

File: CHANGELOG.rst
Comment (robrap) [inline CHANGELOG.rst:24]: Consider adding a note that the Legacy classes are deprecated and will soon be removed.  Since this note would probably only benefit you, and you already know this, you can probably ignore.

**Review by robrap — APPROVED** [review]
Minor nits if you want to update, but this is ready to go.  Monitoring shows we should be ready for 2.0.0.  

Unfortunately, `deprecated_incompatible_legacy_waffle_class` is too new, so we don't have any old data to prove that it ever worked.  It also doesn't have tests.  However, if we assume it works as planned, then we are all set.  I am getting data for `deprecated_compatible_legacy_waffle_class`, so hopefully that is a good enough indicator.

File: CHANGELOG.rst
Comment (regisb) [reply CHANGELOG.rst:28]: Right! I missed those.

**Comment by robrap** [comment]
@regisb: Any reason you didn't merge, or just because you needed to wait for green build?

**Comment by regisb** [comment]
Sorry, I thought you would merge it! I'll make the release now.

**Pull request merged**
